pydicom/values.py

```diff
@@ -525,11 +525,11 @@ def convert_text(
     str or list of str
         The decoded value(s).
     """
-    values = byte_string.split(b'\\')
-    as_strings = [convert_single_string(value, encodings, vr)
-                  for value in values]
+    single_string = convert_single_string(byte_string, encodings, vr)
+    as_strings = single_string.split('\\')
     if len(as_strings) == 1:
         return as_strings[0]
+    as_strings = [value.rstrip('\0 ') for value in as_strings]
```

Comment:
Why is this needed? As far as I know, null bytes are only used in `UI` strings for padding, and the space padding for the last value should already be handled.

Reply:
I'm not familiar with the DICOM standard, but test cases below failed without this line.

https://github.com/pydicom/pydicom/blob/a8be738418dee0a2b93c241fbd5e0bc82f4b8680/pydicom/tests/test_values.py#L114-L122

Reply:
Hm, maybe there has been some invalid DICOM that did this, so the tests have been added. 
Anyway, I'm not sure why these test cases would fail - I will have a closer look tonight or over the weekend.

Reply:
Ok, I had another look and understand it now. The removal of the padding is done in `convert_single_string`, alongside with the validation - which may be a problem, because the validation assumes single values and may fail if the combined string is longer than a single string is allowed to be. I don't have a nice solution right now. We could do the decoding part first, and then do the validation and stripping on each part, e.g. something like:
```py
def convert_text(...):
    def handle_value(v):
        if vr is not None:
            validate_value(
                vr, v, config.settings.reading_validation_mode)
        return v.rstrip('\0 ')

    encodings = encodings or [default_encoding]
    decoded_string = decode_bytes(byte_string, encodings, TEXT_VR_DELIMS)
    values = decoded_string.split("\\")
    as_strings = [handle_value(value) for value in values]
    if len(as_strings) == 1:
        return as_strings[0]
    return MultiValue(str, as_strings,
                      validation_mode=config.settings.reading_validation_mode)
```
I'm not sure about the validation, because it will have to encode the string again to check the length, so this is a bit ugly. I will think about it over the weekend, or maybe you have a better idea meanwhile...

Reply:
Considering `convert_text` is applied to LO, SH, and UC and the maximum lengths for LO and SH are specified in chars not in bytes (and validation for UC is unnecessary?), the code you provided looks perfect to me.

Reply:
Also on a related function `convert_single_string`, decoding of ST (1024 chars max) like below is not supposed to raise error or show warning if my understanding of the value length validation is correct.

```python
encodings = ['iso_ir_126', 'iso_ir_144']
original_string = 'ΔЛ' * 512
bytestring = pydicom.charset.encode_string(original_string, encodings)
pydicom.values.convert_single_string(bytestring, encodings, 'ST')
```
> UserWarning: The value length (4096) exceeds the maximum length of 1024 allowed for VR ST.

Quick fix would be doing validation after decoding but I'm not sure if that could cause some bad side effects or not.

Reply:
You are completely right - this is incorrectly implemented. I think I misread the standard at the time I did this, the check shall indeed always be done after the decoding, both in `convert_text` and `convert_single_string`. This would check the character length instead of the byte length. As far as I can see, there should be no side effects (other than encoding errors will be found before validation errors, which should not be a problem). 

We have also to add a couple of tests or adapt the existing ones.
Additionally, there is a call `validate_value` in `filewriter._encode_and_validate_string`, which can probably be removed completely (and the method call just replaced by a call to `encode_string`), as it assumes that the decoded version shall be re-validated.

How do you want to proceed? If you want, you can make these changes yourself, or I can go ahead and fix this, either in a separate PR, or by adding commits to this PR.

Thanks for pointing this problem out - this has not been noticed so far, and had slipped through the review!

Reply:
Thank YOU for reviewing and everything!

I've committed what I thought was minimally required for this PR. I tried to fix similar/related issues as you suggested but the scope of the relevant code seemed broader than I could handle at the moment. So, I hope you can handle the rest (I don't mind where it would be done; whether in this PR or in a new PR).

FWIW, let me dump here what I thought had to be added/fixed (and gave up actually doing).

## Add unit test for `convert_PN`

Decoding of PN seemed need fixing.

```python
class TestConvertPN:
    def test_valid_PN(self):
        bytestring = b'John^Doe'
        assert 'John^Doe' == convert_PN(bytestring)

        bytestring = b'Yamada^Tarou=\x1b$B;3ED\x1b(B^\x1b$BB@O:'
        assert 'Yamada^Tarou=山田^太郎' == convert_PN(bytestring, ['latin_1', 'iso2022_jp'])

        # 0x5c (\) in multibyte codes
        bytestring = b'Baishaku^Honme=\x1b$BG\\<\\\x1b(B^\x1b$BK\\L\\'
        assert 'Baishaku^Honme=倍尺^本目' == convert_PN(bytestring, ['latin_1', 'iso2022_jp'])

        # 0x3d (=) in multibyte codes
        bytestring = b'Sunatouge^Ayano=\x1b$B:=F=\x1b(B^\x1b$B0=G='
        assert 'Sunatouge^Ayano=砂峠^綾能' == convert_PN(bytestring, ['latin_1', 'iso2022_jp'])
```

## Fix tests for value length

https://github.com/pydicom/pydicom/blob/1db21d627e2e57dcc3aa4ec15666e71267220f62/pydicom/tests/test_dataelem.py#L939-L982

Reply:
Thank you! I added a commit which should fix the issue, though that raises another question (probably outside of this PR).

@darcymason - I just basically removed the write_validation part, as the string validation I thought was needed is not actually needed. I'm not sure if we should add validation to all values on writing instead... otherwise that write validation config does not make much sense. Or maybe this is another thing that shall be considered for 3.0...

Reply:
I've been sorting through my thoughts on validation config and playing with some code snippets.  Hopefully will start a separate discussion on that in the next couple of days.

I'll look at and comment on the specifics here too, hopefully tomorrow, but in general I don't think we should worry too much because I expect it will all be overhauled after the next release.

Reply:
Ok, thanks!

Reply:
I've looked at this again, and I think there still is a role for validation in these (at the Setting stage in discussion #1726), but it should happen before encoding, maybe just checking max length, e.g.  But we can do that in a separate PR (for 2.4) with some initial moves in the direction of pydicom 3.0 where possible.